File: algorithms/python/字母大小写全排列/字母大小写全排列.py
# Source : https://leetcode.cn/problems/letter-case-permutation/description/
# Author : Yanan Wang
# Date   : 2023-09-06

# 该题先画一颗决策树，然后根据决策树，再利用全局变量ans和path即可解答本题
#              当递归到数字时，直接将数字插进去递归下一层
#              当递归到字母时，分两种情况递归：1、不转变大小写  2、转变大小写

class Solution:
    def letterCasePermutation(self, s: str) -> List[str]:
        ans, path = [], []
        def dfs(index = 0):
            if len(s) == index:
                ans.append(''.join(path))
                return
            if s[index].isalpha():
                if s[index].islower():
                    path.append(s[index].upper())
                else:
                    path.append(s[index].lower())
                dfs(index + 1)
                path.pop()
            path.append(s[index])
            dfs(index + 1)
            path.pop()

        dfs()
        return ans

File: algorithms/cpp/字母大小写全排列/字母大小写全排列.cpp
// Source : https://leetcode.cn/problems/letter-case-permutation/description/
// Author : Yanan Wang
// Date   : 2023-09-07

// 本题先画一颗决策树：当递归到数字时，直接添加上递归到下一层，当递归到字母时，转变大小写进行一次递归，不转变大小写再进行一次递归

class Solution {
private:
    vector<string> ans;
    string path;
public:
    void dfs(const string &s, int index) {
        if (s.size() == index) {
            ans.push_back(path);
            return;
        }
        if (isalpha(s[index])) {
            if (isupper(s[index])) {
                path.push_back(tolower(s[index]));
            } else {
                path.push_back(toupper(s[index]));
            }
            dfs(s, index + 1);
            path.pop_back();
        }
        path.push_back(s[index]);
        dfs(s, index + 1);
        path.pop_back();
    }
    vector<string> letterCasePermutation(string s) {
        dfs(s, 0);
        return ans;
    }
};